src/transformStateWithClones.js

```diff
@@ -7,7 +7,27 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  let currentState = { ...state };
+
+  const stateHistory = [];
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      const newState = { ...currentState };
 
+      for (const key of action.keysToRemove) {
+        delete newState[key];
+      }
+      currentState = newState;
+    }
```

Comment:
There should be a `default` case in the `switch` statement to handle any unexpected action types. This is a good practice for error handling and code robustness.